### PR TITLE
Check Groovy syntax with JJB templates

### PR DESCRIPTION
**Pull request by hughsaunders**
This commit extracts groovy from JJB templates so they can be
checked for syntax. Previously JJB templates were only checked for
yaml syntax and JJB semantics so groovy errors could slip in.

This commit also reduces lint time by scanning all groovy files
with a single JVM rather than starting one up for each file.

Connects rcbops/u-suk-dev#1661

### CODE REVIEW

File: rpc_jobs/lem_multi_node_aio.yml [hunk 1]
```diff
@@ -146,14 +146,15 @@
...
-            maas.verify()
-            tempest.tempest("infra1", "deploy1")
-            kibana.kibana("deploy1")
-            holland.holland()
-          }} // hardware node
-          ssh_slave.destroy()
-          maas.entity_cleanup(instance_name: instance_name)
-        }} // cit node
-      }} // timeout
+              maas.prepare(instance_name: instance_name)
```
Comment (hughsaunders) [inline rpc_jobs/lem_multi_node_aio.yml]: This change was required for tests to pass. I only fixed syntax, I think the job still has problems (we shouldn't be running maas stuff on the cit slave).

Comment (git-harry) [inline rpc_jobs/lem_multi_node_aio.yml]: Have you created an issue to track the issues you've found?

**Comment by hughsaunders** [comment]
@git-harry I agree that would be better.  https://docs.openstack.org/infra/jenkins-job-builder/project_pipeline.html

File: rpc_jobs/lem_multi_node_aio.yml (hunk 1 above)
Comment (hughsaunders) [inline rpc_jobs/lem_multi_node_aio.yml]: rcbops/u-suk-dev#1666

**Comment by hughsaunders** [comment]
In order to use that, we would need to update to JJB 2.0.0.0b2:

```
$ git tag --contains ":/Add Pipeline job type"
2.0.0.0b2
```

I would suggest that we wait till its released, except that the [version we currently use](https://github.com/rcbops/rpc-gating/blob/master/constraints.txt#L15) is 2.0.0.0b1.

**Comment by hughsaunders** [comment]
In order to use that, we would need to update to JJB 2.0.0.0b2:

```
$ git tag --contains ":/Add Pipeline job type"
2.0.0.0b2
```

I would suggest that we wait till its released, except that the [version we currently use](https://github.com/rcbops/rpc-gating/blob/master/constraints.txt#L15) is 2.0.0.0b1.

**Comment by hughsaunders** [comment]
Unlike inline scripts, pipeline scripts from scm are [not trusted](https://wiki.jenkins-ci.org/display/JENKINS/Script+Security+Plugin#ScriptSecurityPlugin-User%E2%80%99sguide), so many functions must be [approved for in-process execution](https://rpc.jenkins.cit.rackspace.net/scriptApproval/). Unfortunately some of the functions cannot be approved for example:

```
org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: unclassified method java.util.LinkedHashMap step
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:113)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:149)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:146)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:16)
	at Script1.conditionalStep(Script1.groovy:228)
	at Script2.delPubCloudSlave(Script2.groovy:95)
	at Script2.runonpubcloud(Script2.groovy:120)
	at WorkflowScript.run(WorkflowScript:26)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:57)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:109)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixName(FunctionCallBlock.java:77)
	at sun.reflect.GeneratedMethodAccessor284.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
	at com.cloudbees.groovy.cps.Next.step(Next.java:74)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:154)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:18)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:33)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:30)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:108)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:30)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:165)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:328)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$100(CpsThreadGroup.java:80)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:240)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:228)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Finished: FAILURE
```


There is an [open issue](https://issues.jenkins-ci.org/browse/JENKINS-28178) for allowing scripts from scm to be trusted. 

One option would be to use the [permissive script security plugin](https://wiki.jenkins-ci.org/display/JENKINS/Permissive+Script+Security+Plugin), but that isn't recommended. 

So using pipeline-scm is not as simple/clean as it appears :/

**Comment by hughsaunders** [comment]
Reverted to previous version due to problems with script security when using pipeline-scm.

File: lint.sh [hunk 2]
```diff
@@ -29,23 +29,45 @@ check_jjb(){
     || { echo "JJB Syntax fail"; rc=1; }
 }
 
+# This pulls job.dsl from jjb templates so they can be checked for groovy syntax
+extract_groovy_from_jjb(){
+  mkdir -p tmp_groovy
+  pushd rpc_jobs
+  for i in *.yml
+  do
+    python <<EOF
```
Comment (hughsaunders) [inline lint.sh]: I do see the irony of adding an embeded script in a PR for extracting embedded scripts :/ I'll move it out.

File: lint.sh [hunk 3]
```diff
@@ -29,23 +29,45 @@ check_jjb(){
...
+import yaml
+with open("${i}",'r') as inf:
+  jjb = yaml.load(inf)
```
Comment (hughsaunders) [inline lint.sh]: done

File: lint.sh [hunk 4]
```diff
@@ -29,23 +29,45 @@ check_jjb(){
...
+  key = item.keys()[0]
+  if key in ("job", "job-template") and 'dsl' in item[key]:
+    outfile="${i}".split('.')[0]
+    with open("../tmp_groovy/{outfile}-{item}.groovy".format(
+        outfile=outfile, item=item[key]['name']), "w") as outf:
+      outf.write(item[key]['dsl'])
+EOF
+  done
+  popd
+  sed -i bk -e 's/{{/{/g' -e 's/}}/}/g' tmp_groovy/*
```
Comment (hughsaunders) [inline lint.sh]: done

File: scripts/syntax.groovy [hunk 5]
```diff
@@ -0,0 +1,7 @@
+ClassLoader gcl = new GroovyClassLoader()
+for (scriptfile in args) {
+  print("${scriptfile} ")
+  String script = new File(scriptfile).text
+  Class c = gcl.parseClass(script)
```
Comment (hughsaunders) [inline scripts/syntax.groovy]: done

**Comment by hughsaunders** [comment]
@git-harry thanks for the review, I think I've addressed your comments.

File: lint.sh (hunk 2 above)
Comment (hughsaunders) [inline lint.sh]: done

File: scripts/syntax.groovy [hunk 6]
```diff
@@ -0,0 +1,15 @@
+ClassLoader gcl = new GroovyClassLoader()
+rc=0
+for (scriptfile in args) {
+  try{
+    print("${scriptfile} ")
+    String script = new File(scriptfile).text
+    Class c = gcl.parseClass(script)
+  } catch (e){
+    print("[PARSE ERROR]\n")
```
Comment (git-harry) [inline scripts/syntax.groovy]: What about printing the error as well, e.g
```groovy
    print("[PARSE ERROR]\n$e\n")
```

Comment (hughsaunders) [inline scripts/syntax.groovy]: done

**Comment by hughsaunders** [comment]
@git-harry addressed last request.

**Pull request merged**